diff --git a/Makefile b/Makefile
index 231ecd0..9a25c3b 100644
--- a/Makefile
+++ b/Makefile
@@ -89,10 +89,10 @@ export MODLIB
 # standard CFLAGS
 #
 
-CPPFLAGS := -D__KERNEL__ -I$(HPATH)
+CPPFLAGS := -D__KERNEL__ -I$(HPATH) -g
 
 CFLAGS := $(CPPFLAGS) -Wall -Wstrict-prototypes -Wno-trigraphs -O2 \
-	  -fno-strict-aliasing -fno-common
+	  -fno-strict-aliasing -fno-common -g
 ifndef CONFIG_FRAME_POINTER
 CFLAGS += -fomit-frame-pointer
 endif
