diff --git a/scripts/Configure b/scripts/Configure
index cdc48a8..5d003cd 100644
--- a/scripts/Configure
+++ b/scripts/Configure
@@ -539,13 +539,15 @@ if [ -f .config ]; then
   DEFAULTS=.config
 fi
 
+DEFAULTS=$(realpath $DEFAULTS)
+
 if [ -f $DEFAULTS ]; then
   echo "#"
   echo "# Using defaults found in" $DEFAULTS
   echo "#"
   . $DEFAULTS
   sed -e 's/# \(CONFIG_[^ ]*\) is not.*/\1=n/' <$DEFAULTS >.config-is-not.$$
-  . .config-is-not.$$
+  . $PWD/.config-is-not.$$
   rm .config-is-not.$$
 else
   echo "#"
